source/secure_tunneling.c

```diff
@@ -423,7 +435,28 @@ static int s_secure_tunneling_send_data(struct aws_secure_tunnel *secure_tunnel,
         AWS_LOGF_ERROR(AWS_LS_IOTDEVICE_SECURE_TUNNELING, "Invalid Stream Id");
         return AWS_ERROR_IOTDEVICE_SECUTRE_TUNNELING_INVALID_STREAM;
     }
-    return s_secure_tunneling_send(secure_tunnel, data, DATA);
+    struct aws_byte_cursor new_data;
+    new_data.ptr = data->ptr;
+    new_data.len = data->len;
+    while (new_data.len) {
+        size_t bytes_max = new_data.len;
+        size_t amount_to_send = bytes_max < AWS_IOT_ST_SPLIT_MESSAGE_SIZE ? bytes_max : AWS_IOT_ST_SPLIT_MESSAGE_SIZE;
+
+        struct aws_byte_cursor send_cursor = aws_byte_cursor_advance(&new_data, amount_to_send);
+        if (send_cursor.len) {
+            secure_tunnel->data_callback_flag = false;
+            if (s_secure_tunneling_send(secure_tunnel, &send_cursor, DATA) != AWS_OP_SUCCESS) {
+                AWS_LOGF_ERROR(AWS_LS_IOTDEVICE_SECURE_TUNNELING, "Failure writing data to out_buf");
+                return AWS_OP_ERR;
+            }
+        }
+        aws_mutex_lock(&secure_tunnel->send_data_mutex);
+        while (secure_tunnel->data_callback_flag == false) {
```

Comment:
use the aws_condition_variable_wait_pred  where the predicate checks the value of the bool.  Removes the need for you to have a while loop